File: MPC NMPC Agus/Kodingan_MPC_NMPC/MPC/MPC_Biasa_Lurus.m
clc; clear; close all;

%% Parameter Kapal
L = 101.07;         % Panjang Kapal (m)
B = 14;             % Lebar Kapal (m)
T = 3.7;            % Draft Kapal (m)
m = 2423*1e3;       % Massa Kapal (kg)
u_0 = 15.4;         % Kecepatan surge Kapal (m/s)
C_B = 0.65;         % Koefisien Blok
x_G = 5.25;         % Pusat massa sumbu-x
A_delta = 5.7224;   % Luas rudder (m^2)
rho = 1024;         % Massa jenis air laut (kg/m^3)
r = 0.156*L;        % Kisaran jari-jari girisa

%% Koefisien Hidrodinamika
Y_v_dot = -(1 + 0.16*C_B*B/T - 5.1*(B/L)^2)*pi*(T/L)^2;
Y_r_dot = -(0.67*(B/L) - 0.0033*(B/T)^2)*pi*(T/L)^2;
N_v_dot = -(1.1*B/L - 0.041*B/T)*pi*(T/L)^2;
N_r_dot = -((1/12) + 0.017*C_B*B/T - 0.33*B/L)*pi*(T/L)^2;
Y_v = -(1 + 0.4*C_B*B/T)*pi*(T/L)^2;
Y_r = -(-0.5 + 2.2*B/L - 0.08*B/T)*pi*(T/L)^2;
N_v = -(0.5 + 2.4*T/L)*pi*(T/L)^2;
N_r = -(0.25 + 0.039*B/T - 0.56*B/L)*pi*(T/L)^2;
Y_delta = (rho*pi*A_delta)/(4*L*T);
N_delta = -0.5*Y_delta;

m_nd = 2*m/(rho*L^3);
x_G_nd = x_G/L;
I_z_nd = 1.2392*10^(-4);
u_0_nd = 1;

%% Model Gerak Kapal (Linier)
M = [m_nd-Y_v_dot , m_nd*x_G_nd-Y_r_dot ;
    m_nd*x_G_nd - N_v_dot , I_z_nd - N_r_dot ];

a11 = ((I_z_nd - N_r_dot)*Y_v - (m_nd*x_G_nd - Y_r_dot)*N_v)/det(M);
a12 = ((I_z_nd - N_r_dot)*(Y_r - m_nd*u_0_nd) - (m_nd*x_G_nd - Y_r_dot)*(N_r - m_nd*x_G_nd*u_0_nd))/det(M);
a21 = ((m_nd - Y_v_dot)*N_v - (m_nd*x_G_nd - N_v_dot)*Y_v)/det(M);
a22 = ((m_nd - Y_v_dot)*(N_r - m_nd*x_G_nd*u_0_nd) - (m_nd*x_G_nd - N_v_dot)*(Y_r - m_nd*u_0_nd))/det(M);

A_sys = [a11 , a12 ; a21 , a22 ];
B_sys = [0.01 ; 1];

% State order: [v; r; x; y; psi]
A_lin = zeros(5,5);
A_lin(1:2, 1:2) = A_sys;
A_lin(4, 1) = 1;        
A_lin(4, 5) = u_0_nd;   
A_lin(5, 2) = 1;        

B_lin = zeros(5,1);
B_lin(1:2) = B_sys;

D_affine = zeros(5,1);
D_affine(3) = u_0_nd;   

%% Setup LMPC Konvensional (Beda Hingga)
Tp = 60;            % Horizon waktu prediksi (s)
Np = 60;            % Jumlah langkah prediksi (menggunakan dt = 1 detik)
dt_pred = Tp / Np;  % Interval waktu prediksi diskrit
T_sim = 1;          % Langkah waktu simulasi (s)
T_sim_total = 180;  % Waktu simulasi total (s)

% Diskritisasi Beda Hingga (Euler Forward) ke Non-Dimensional
dt_nd = dt_pred * (u_0 / L);
Ad = eye(5) + A_lin * dt_nd;
Bd = B_lin * dt_nd;
Dd = D_affine * dt_nd;

% Bobot
Q = diag([50, 50, 50]);        
Q_full = diag([0, 0, Q(1,1), Q(2,2), Q(3,3)]); 
R = 100;                      

% Batasan state
r_limit = 0.0932;
r_limit_nd = r_limit * (L / u_0);
lb_state_single = [-Inf; -r_limit_nd; -Inf; -Inf; -Inf];
ub_state_single = [ Inf;  r_limit_nd;  Inf;  Inf;  Inf];

% Batasan input
u_limit = deg2rad(35);
lb_input_single = -u_limit;
ub_input_single =  u_limit;

% Batasan perubahan input (dikalikan dt_pred karena rate adalah per detik)
u_rate_limit = deg2rad(5) * dt_pred;

lb = [repmat(lb_state_single, Np+1, 1); repmat(lb_input_single, Np, 1)];
ub = [repmat(ub_state_single, Np+1, 1); repmat(ub_input_single, Np, 1)];

%% Membangun Matriks QP LMPC Konvensional
% 1. Matriks Hessian (H) 
H_s = kron(eye(Np+1), Q_full);
H_u = kron(eye(Np), R);
H = blkdiag(H_s, H_u);
H = (H + H') / 2; 
H = H + 1e-6 * eye(size(H)); 

% 2. Kendala Persamaan Dinamika Kapal: s(k+1) = Ad*s(k) + Bd*u(k) + Dd
A_eq_dyn_s = kron(eye(Np+1), eye(5));
for i = 1:Np
    A_eq_dyn_s(i*5+1:(i+1)*5, (i-1)*5+1:i*5) = -Ad;
end
A_eq_dyn_u = zeros(5*(Np+1), Np);
for i = 1:Np
    A_eq_dyn_u(i*5+1:(i+1)*5, i) = -Bd;
end
A_eq = [A_eq_dyn_s, A_eq_dyn_u];

b_eq_base = zeros(5*(Np+1), 1);
for i = 1:Np
    b_eq_base(i*5+1:(i+1)*5) = Dd;
end
% Note: Elemen 1-5 dari b_eq_base akan diisi dengan x0 saat di dalam loop.

% 3. Kendala Pertidaksamaan Laju Input (delta u)
A_ineq_rate = zeros(Np, Np);
A_ineq_rate(1,1) = 1;
for i = 2:Np
    A_ineq_rate(i, i) = 1;
    A_ineq_rate(i, i-1) = -1;
end
A_ineq_u = [A_ineq_rate; -A_ineq_rate];
A_ineq = [zeros(2*Np, 5*(Np+1)), A_ineq_u];

%% Simulasi LMPC Konvensional
x0 = [0; 0; 0; 500; 0];         
x0_nd = dimensional_to_nondimensional(x0, L, u_0); 
h_ref = [0; 0; 0];              
history_state_nd = [];
history_input = [];
input_initial = 0;                               

options = optimoptions('quadprog', 'Display', 'none','MaxIterations', 2000);
fprintf('Memulai simulasi Linear MPC (Konvensional/Euler)...\n');
total_timer = tic;

for t = 0:T_sim:T_sim_total
    
    % --- Update Vektor Referensi ---
    ref_future = zeros(5, Np+1); 
    for k = 1:Np+1
        t_predict_sekat = t + (k-1) * dt_pred; 
        ref_future(3, k) = (h_ref(1) + t_predict_sekat * u_0) / L; 
        ref_future(4, k) = h_ref(2) / L;                           
        ref_future(5, k) = h_ref(3);                               
    end
    s_ref_flat = ref_future(:);
    
    % --- Update Vektor Gradien (f) ---
    f_s = -H_s * s_ref_flat;
    f_u = zeros(Np, 1);
    f = [f_s; f_u];
    
    % --- Update Batasan Kondisi Awal ---
    b_eq = b_eq_base;
    b_eq(1:5) = x0_nd; % Mengunci s_0 agar sama dengan kondisi aktual
    
    % --- Update Batasan Laju Input Awal ---
    b_ineq_limit1 = u_rate_limit * ones(Np, 1);
    b_ineq_limit1(1) = u_rate_limit + input_initial; % Batas atas u_0
    
    b_ineq_limit2 = u_rate_limit * ones(Np, 1);
    b_ineq_limit2(1) = u_rate_limit - input_initial; % Batas bawah u_0
    
    b_ineq = [b_ineq_limit1; b_ineq_limit2];
    
    % --- Optimasi QP ---
    [z_opt, ~, exitflag] = quadprog(H, f, A_ineq, b_ineq, A_eq, b_eq, lb, ub, [], options);
    
    if exitflag ~= 1
        warning('Quadprog gagal konvergen pada t = %d. Exitflag: %d', t, exitflag);
    end
    
    % Ambil input pertama (u_0) dari solusi optimal. Terletak setelah vektor state s.
    u_apply = z_opt(5*(Np+1) + 1); 
    input_initial = u_apply;          
    
    % Simpan hasil
    history_state_nd = [history_state_nd, x0_nd];
    history_input = [history_input, u_apply];
    
    % Simulasi dinamika linier kapal (Runge-Kutta 4)
    dt_nd_sim = T_sim * u_0 / L; 
    k1 = linear_ship_dynamics(x0_nd, u_apply, A_lin, B_lin, D_affine);
    k2 = linear_ship_dynamics(x0_nd + 0.5*dt_nd_sim*k1, u_apply, A_lin, B_lin, D_affine);
    k3 = linear_ship_dynamics(x0_nd + 0.5*dt_nd_sim*k2, u_apply, A_lin, B_lin, D_affine);
    k4 = linear_ship_dynamics(x0_nd + dt_nd_sim*k3, u_apply, A_lin, B_lin, D_affine);
    x0_nd = x0_nd + (dt_nd_sim/6) * (k1 + 2*k2 + 2*k3 + k4);
    
    if mod(t, 20) == 0
        disp(['Waktu Simulasi: ', num2str(t), ' s']);
    end
end
waktu_total = toc(total_timer);
fprintf('SIMULASI SELESAI\n');
fprintf('Total Waktu Komputasi MPC Konvensional: %.4f detik\n', waktu_total);

history_state_dim = zeros(size(history_state_nd));
for i = 1:size(history_state_nd, 2)
    history_state_dim(:, i) = nondimensional_to_dimensional(history_state_nd(:, i), L, u_0);
end

%% Plot Hasil (Lengkap dengan Verifikasi Batasan)
figure('Name', 'Hasil Simulasi LMPC Konvensional');
time_vector = 0:T_sim:T_sim_total; 

% Plot 1: Trajektori
subplot(2,1,1);
plot(history_state_dim(3,:), history_state_dim(4,:), 'b-', 'LineWidth', 2); hold on;
plot([min(history_state_dim(3,:)) max(history_state_dim(3,:))], [0 0], 'r--', 'LineWidth', 1.5);
xlabel('Posisi X'); ylabel('Posisi Y');
title('Trajektori Kapal vs Referensi');
legend('Kapal', 'Referensi'); grid on; axis equal; 

% Plot 2: Heading
subplot(2,1,2);
plot(time_vector, rad2deg(history_state_dim(5,:)), 'b-', 'LineWidth', 2); hold on;
y_ref_heading = rad2deg(h_ref(3)) * ones(size(time_vector));
plot(time_vector, y_ref_heading, 'r--', 'LineWidth', 1.5);
xlabel('Waktu (s)'); ylabel('Heading (derajat)');
title('Sudut Haluan (Yaw)');
legend('Actual', 'Reff'); grid on;

%% Plot Verifikasi Batasan (Constraints)
figure('Name', 'Verifikasi Batasan (Constraints)');

% Plot 1: Yaw Rate
subplot(3,1,1);
plot(time_vector, rad2deg(history_state_dim(2, :)), 'b-', 'LineWidth', 2); hold on;
yline(rad2deg(r_limit), 'r--', 'LineWidth', 1.5, 'Label', 'Max Limit');
yline(-rad2deg(r_limit), 'r--', 'LineWidth', 1.5, 'Label', 'Min Limit');
xlabel('Waktu (s)'); ylabel('Yaw Rate (deg/s)');
title('Batasan Yaw Rate (r)'); legend('Actual r', 'Limit'); grid on;
ylim([-rad2deg(r_limit)*1.5, rad2deg(r_limit)*1.5]);

% Plot 2: Rudder Input
subplot(3,1,2);
plot(time_vector, rad2deg(history_input), 'g-', 'LineWidth', 2); hold on;
yline(rad2deg(u_limit), 'r--', 'LineWidth', 1.5);
yline(-rad2deg(u_limit), 'r--', 'LineWidth', 1.5);
xlabel('Waktu (s)'); ylabel('Sudut Rudder (derajat)');
title('Kontrol Input'); legend('Rudder Angle', 'Limit'); grid on; 
ylim([-40 40]);

% Plot 3: Rudder Rate
subplot(3,1,3);
u_actual_deg = rad2deg(history_input); 
% Karena T_sim = 1 detik, perubahan input langsung mencerminkan deg/s
u_rate_calc = [0, diff(u_actual_deg)] / T_sim; 
plot(time_vector, u_rate_calc, 'm-', 'LineWidth', 2); hold on;

% Batas laju putar kemudi fisik adalah 5 derajat/detik
yline(5, 'r--', 'LineWidth', 1.5, 'Label', 'Max Rate');
yline(-5, 'r--', 'LineWidth', 1.5, 'Label', 'Min Rate');
xlabel('Waktu (s)'); ylabel('Rudder Rate (deg/s)');
title('Verifikasi Batasan Perubahan Sudut Rudder (derajat)'); 
legend('Perubahan Input Rudder', 'Limit'); grid on;
ylim([-15 15]);

%% Fungsi Dinamika Linier Kapal
function s_dot = linear_ship_dynamics(s, u, A_lin, B_lin, D_affine)
     s_dot = A_lin * s + B_lin * u + D_affine;
end

%% Fungsi Konversi Nondimensional
function x_nd = dimensional_to_nondimensional(x_dim, L, u_0)
     x_nd = [x_dim(1)/u_0; x_dim(2)*L/u_0; x_dim(3)/L; x_dim(4)/L; x_dim(5)];
end
function x_dim = nondimensional_to_dimensional(x_nd, L, u_0)
     x_dim = [x_nd(1)*u_0; x_nd(2)*u_0/L; x_nd(3)*L; x_nd(4)*L; x_nd(5)];
end

File: MPC NMPC Agus/Kodingan_MPC_NMPC/MPC/Python/mpc_biasa_lurus.py
"""
Simulasi Linear MPC konvensional — port Python dari MPC/MATLAB/MPC_Biasa_Lurus.m

Jalankan:
    python mpc_biasa_lurus.py
"""

from __future__ import annotations

import time
import warnings

import matplotlib.pyplot as plt
import numpy as np

from mpc_model import (
  MPCConfig,
  ShipParams,
  build_linear_ship_model,
  dimensional_to_nondimensional,
  nondimensional_to_dimensional,
  rk4_step_linear,
)
from mpc_qp import (
  build_mpc_qp_problem,
  build_reference_vector,
  equality_rhs,
  gradient,
  inequality_rhs,
  solve_mpc_step,
)


def run_simulation(
  plot: bool = True,
  show_plots: bool = True,
  log_step_time: bool = True,
) -> dict:
  """Jalankan simulasi LMPC penuh (default 180 detik, Np=60)."""
  ship = ShipParams()
  cfg = MPCConfig()
  l_ship, u_0 = ship.L, ship.u_0

  model = build_linear_ship_model(ship, cfg)
  qp = build_mpc_qp_problem(model, cfg.np_steps, cfg.dt_pred)

  x0_dim = np.array([0.0, 0.0, 0.0, 500.0, 0.0])
  x0_nd = dimensional_to_nondimensional(x0_dim, l_ship, u_0)
  h_ref = np.array([0.0, 0.0, 0.0])

  history_state_nd: list[np.ndarray] = []
  history_input: list[float] = []
  step_times_s: list[float] = []
  qp_iters: list[int] = []
  qp_status: list[str] = []

  u_prev = 0.0
  x_nd = x0_nd.copy()

  num_steps = int(cfg.t_sim_total / cfg.t_sim) + 1
  dt_nd_sim = cfg.t_sim * u_0 / l_ship

  print("Memulai simulasi Linear MPC (Konvensional/Euler)...")
  if log_step_time:
    print(
      f"{'t (s)':>8}  {'waktu (ms)':>10}  {'iter':>6}  {'status':>6}"
    )
    print("-" * 40)

  t_start = time.perf_counter()

  for step in range(num_steps):
    t = step * cfg.t_sim
    step_start = time.perf_counter()

    s_ref_flat = build_reference_vector(qp, t, h_ref, u_0, l_ship)
    f = gradient(qp, s_ref_flat)
    b_eq = equality_rhs(qp, x_nd)
    b_ineq = inequality_rhs(qp, u_prev)

    z_opt, nit, status = solve_mpc_step(qp, f, b_eq, b_ineq)
    if status != "ok":
      warnings.warn(
        f"QP gagal pada t={t} s (iter={nit}): {status}",
        RuntimeWarning,
        stacklevel=2,
      )

    u_apply = float(z_opt[qp.idx_u0])
    u_prev = u_apply

    history_state_nd.append(x_nd.copy())
    history_input.append(u_apply)

    x_nd = rk4_step_linear(
      x_nd,
      u_apply,
      dt_nd_sim,
      model.a_lin,
      model.b_lin,
      model.d_affine,
    )

    step_elapsed = time.perf_counter() - step_start
    step_times_s.append(step_elapsed)
    qp_iters.append(nit)
    qp_status.append(status)

    if log_step_time and (step % 20 == 0 or step == num_steps - 1):
      print(
        f"{t:8.1f}  {step_elapsed * 1000.0:10.2f}  {nit:6d}  {status:>6}"
      )

  elapsed = time.perf_counter() - t_start
  step_times_arr = np.array(step_times_s)

  print("SIMULASI SELESAI")
  print(f"Total Waktu Komputasi MPC Konvensional: {elapsed:.4f} detik")
  if len(step_times_arr) > 0:
    print(
      "Waktu per langkah t - "
      f"min: {step_times_arr.min() * 1000.0:.2f} ms, "
      f"max: {step_times_arr.max() * 1000.0:.2f} ms, "
      f"rata-rata: {step_times_arr.mean() * 1000.0:.2f} ms"
    )

  hist_nd = np.column_stack(history_state_nd)
  hist_dim = np.column_stack([
    nondimensional_to_dimensional(hist_nd[:, i], l_ship, u_0)
    for i in range(hist_nd.shape[1])
  ])
  history_input_arr = np.array(history_input)

  time_vector = np.arange(num_steps, dtype=float) * cfg.t_sim

  x_ref_full = h_ref[0] + time_vector * u_0
  y_ref_full = np.full_like(time_vector, h_ref[1])
  psi_ref_full = np.full_like(time_vector, h_ref[2])

  rmse_x = float(np.sqrt(np.mean((hist_dim[2, :] - x_ref_full) ** 2)))
  rmse_y = float(np.sqrt(np.mean((hist_dim[3, :] - y_ref_full) ** 2)))
  rmse_psi = float(np.sqrt(np.mean((hist_dim[4, :] - psi_ref_full) ** 2)))

  print("\n Hasil Perhitungan RMSE ")
  print(f"RMSE X   : {rmse_x:.4f} meter")
  print(f"RMSE Y   : {rmse_y:.4f} meter")
  print(f"RMSE Psi : {rmse_psi:.4f} rad ({np.rad2deg(rmse_psi):.4f} derajat)\n")

  if plot:
    _plot_results(
      hist_dim,
      history_input_arr,
      time_vector,
      h_ref,
      cfg,
      model,
      show=show_plots,
    )

  return {
    "hist_dim": hist_dim,
    "history_input": history_input_arr,
    "time_vector": time_vector,
    "rmse_x": rmse_x,
    "rmse_y": rmse_y,
    "rmse_psi": rmse_psi,
    "elapsed_s": elapsed,
    "step_times_s": step_times_arr,
    "qp_iters": qp_iters,
    "qp_status": qp_status,
  }


def _plot_results(
  hist_dim: np.ndarray,
  history_input: np.ndarray,
  time_vector: np.ndarray,
  h_ref: np.ndarray,
  cfg: MPCConfig,
  model,
  show: bool = True,
) -> None:
  """Plot trajektori, heading, dan verifikasi batasan (sama seperti MATLAB)."""
  u_limit_deg = cfg.u_limit_deg
  r_limit = cfg.r_limit
  t_sim = cfg.t_sim

  fig1, axes = plt.subplots(2, 1, figsize=(10, 8))
  fig1.canvas.manager.set_window_title("Hasil Simulasi LMPC Konvensional")

  ax1 = axes[0]
  ax1.plot(hist_dim[2, :], hist_dim[3, :], "b-", linewidth=2, label="Kapal")
  x_min, x_max = hist_dim[2, :].min(), hist_dim[2, :].max()
  ax1.plot([x_min, x_max], [0, 0], "r--", linewidth=1.5, label="Referensi")
  ax1.set_xlabel("Posisi X")
  ax1.set_ylabel("Posisi Y")
  ax1.set_title("Trajektori Kapal vs Referensi")
  ax1.legend()
  ax1.grid(True)
  ax1.axis("equal")

  ax2 = axes[1]
  ax2.plot(time_vector, np.rad2deg(hist_dim[4, :]), "b-", linewidth=2, label="Actual")
  ax2.plot(
    time_vector,
    np.full_like(time_vector, np.rad2deg(h_ref[2])),
    "r--",
    linewidth=1.5,
    label="Reff",
  )
  ax2.set_xlabel("Waktu (s)")
  ax2.set_ylabel("Heading (derajat)")
  ax2.set_title("Sudut Haluan (Yaw)")
  ax2.legend()
  ax2.grid(True)

  fig2, axes2 = plt.subplots(3, 1, figsize=(10, 10))
  fig2.canvas.manager.set_window_title("Verifikasi Batasan (Constraints)")

  ax_r = axes2[0]
  ax_r.plot(time_vector, np.rad2deg(hist_dim[1, :]), "b-", linewidth=2, label="Actual r")
  ax_r.axhline(np.rad2deg(r_limit), color="r", linestyle="--", linewidth=1.5, label="Max Limit")
  ax_r.axhline(-np.rad2deg(r_limit), color="r", linestyle="--", linewidth=1.5, label="Min Limit")
  ax_r.set_xlabel("Waktu (s)")
  ax_r.set_ylabel("Yaw Rate (deg/s)")
  ax_r.set_title("Batasan Yaw Rate (r)")
  ax_r.legend()
  ax_r.grid(True)
  ax_r.set_ylim(-np.rad2deg(r_limit) * 1.5, np.rad2deg(r_limit) * 1.5)

  ax_u = axes2[1]
  ax_u.plot(time_vector[: len(history_input)], np.rad2deg(history_input), "g-", linewidth=2)
  ax_u.axhline(u_limit_deg, color="r", linestyle="--", linewidth=1.5)
  ax_u.axhline(-u_limit_deg, color="r", linestyle="--", linewidth=1.5)
  ax_u.set_xlabel("Waktu (s)")
  ax_u.set_ylabel("Sudut Rudder (derajat)")
  ax_u.set_title("Kontrol Input")
  ax_u.grid(True)
  ax_u.set_ylim(-40, 40)

  ax_rate = axes2[2]
  u_deg = np.rad2deg(history_input)
  u_rate = np.concatenate([[0.0], np.diff(u_deg)]) / t_sim
  ax_rate.plot(time_vector[: len(u_rate)], u_rate, "m-", linewidth=2)
  ax_rate.axhline(cfg.u_rate_limit_deg, color="r", linestyle="--", linewidth=1.5, label="Max Rate")
  ax_rate.axhline(-cfg.u_rate_limit_deg, color="r", linestyle="--", linewidth=1.5, label="Min Rate")
  ax_rate.set_xlabel("Waktu (s)")
  ax_rate.set_ylabel("Rudder Rate (deg/s)")
  ax_rate.set_title("Verifikasi Batasan Perubahan Sudut Rudder (derajat)")
  ax_rate.legend()
  ax_rate.grid(True)
  ax_rate.set_ylim(-15, 15)

  plt.tight_layout()
  if show:
    plt.show()


if __name__ == "__main__":
  run_simulation()
